WalletWasabi/Blockchain/BlockFilters/FilterLinkedList.cs

```diff
@@ -0,0 +1,87 @@
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using WalletWasabi.Backend.Models;
+
+namespace WalletWasabi.Blockchain.BlockFilters;
+
+public class FilterLinkedList : IEnumerable<FilterModel>
+{
+	record Node(FilterModel Filter, Node? Next)
+	{
+		public int Count { get; } = (Next?.Count ?? 0) + 1;
+	}
+
+	private Node? _head;
+	private readonly object _syncObject = new();
+
+	public void Add(FilterModel filter)
+	{
+		lock (_syncObject)
+		{
+			_head = new Node(filter, _head);
+		}
+	}
+
+	public void RemoveLast()
+	{
+		lock (_syncObject)
+		{
+			if (_head is not { } nonNullHead)
+			{
+				throw new InvalidOperationException("The linked list is empty.");
+			}
+
+			_head = nonNullHead.Next;
+		}
+	}
+
+	public int Count
+	{
+		get
+		{
+			lock (_syncObject)
+			{
+				return _head switch
+				{
+					null => 0,
+					{ } nonNullHead => nonNullHead.Count
+				};
+			}
+		}
+	}
+
+	public IEnumerator<FilterModel> GetEnumerator()
```

Comment:
Do we need this? Does `IEnumerable<FilterModel>` required?

We only enumerate to find `bestKnownBlockHash` which is not the most performant way of doing that. This class could serve that purpose better because `GetFilterLinesExcluding` is called every time when sync request. For example some kind of Dictionary or Binary search. So in the end enumeration is not required. If we go custom we should really tailor-made stuff.

Reply:
Oh BTW I did not check how tests are using this so we might need enumeration there.

Reply:
We needed it because I wanted to fix a problem **without refactoring the algorithm** and avoiding unnecessary locking.

Using a dictionary can work well but you have to lock because it iterates the internal buckets which are not thread safe. A Binary search requires to sort the list, what needs to lock.

Anyway, the context of this PR changed because now it is not an urgency so, yes, we can think is something different.